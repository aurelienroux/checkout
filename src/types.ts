export interface product {
  id: string
  name: string
  image: string
  variants: variant[]
  description: string
  defaultVariantId: string
}

export interface variant {
  id: string
  name: string
  price: number
  attributes: {
    name: string
    value: string
  }[]
}

export interface updateAttribute {
  attribute: string
  value: string
}

export interface productToAdd {
  description: string
  image: string
  name: string
  price: number
  productId: string
  variantId: string
}

export interface productInCart {
  description: string
  price: number
  productId: string
  qty: number
  variantId: string
}

export enum attributeEnum {
  SIZE = 'Size',
  FABRIC = 'Fabric'
}

export enum cartActionEnum {
  ADD = 'add',
  REMOVE = 'remove'
}
