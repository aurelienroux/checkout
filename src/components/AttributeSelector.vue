<template>
  <div class="selector">
    <span class="selector__name">
      {{ attribute }}
    </span>
    <button
      data-cy="btn-variant"
      class="button button--grey"
      :class="{ 'button--selected': value === selectedValue }"
      v-for="(value, index) in attrOptions"
      :key="index"
      @click="updateAttribute({ attribute, value })"
    >
      {{ value }}
    </button>
  </div>
</template>

<script lang="ts">
import Vue from 'vue'
import { updateAttribute } from '@/types'

export default Vue.extend({
  name: 'AttributeSelector',
  props: {
    attribute: {
      type: String,
      required: true
    },
    attrOptions: {
      type: Array,
      required: true
    },
    selectedValue: {
      type: String,
      required: true
    }
  },
  methods: {
    updateAttribute(attributePayload: updateAttribute) {
      this.$emit('updateAttribute', attributePayload)
    }
  }
})
</script>

<style lang="scss" scoped>
.selector {
  align-items: center;
  display: flex;
  margin-bottom: 1rem;

  &__name {
    color: $text-color;
    flex: 1;
    font-size: 1.1rem;
  }
}
</style>
