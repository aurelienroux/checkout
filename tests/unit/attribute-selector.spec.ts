import { shallowMount } from '@vue/test-utils'

import AttributeSelector from '@/components/AttributeSelector.vue'
import { attributeEnum } from '@/types'

const updateAttribute = jest.fn()

const wrapper = shallowMount(AttributeSelector, {
  propsData: {
    attribute: attributeEnum.SIZE,
    attrOptions: ['36', '38'],
    selectedValue: '36'
  },
  methods: {
    updateAttribute
  }
})

describe('AttributeSelector component', () => {
  it('renders without crashing', () => {
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.findAll('.button').length).toBe(2)
  })

  it('tests if only first button is selected', () => {
    expect(wrapper.findAll('.button').at(0).classes('button--selected')).toBe(true)
    expect(wrapper.findAll('.button').at(1).classes('button--selected')).toBe(false)
  })

  it('triggers updateAttribute method', async () => {
    await wrapper.find('.button').trigger('click')
    expect(updateAttribute).toHaveBeenCalled()
  })
})
