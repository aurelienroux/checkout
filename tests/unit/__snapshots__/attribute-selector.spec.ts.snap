// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AttributeSelector component renders without crashing 1`] = `
<div class="selector"><span class="selector__name">
    Size
  </span> <button data-cy="btn-variant" class="button button--grey button--selected">
    36
  </button><button data-cy="btn-variant" class="button button--grey">
    38
  </button></div>
`;
